### PR TITLE
Add attack functionality to game.  

### PR DESCRIPTION
**Pull request by xMestas**
Attacks can be performed on a board.  Currently working correctly for invalid attacks, hits, misses, and sinks.  Implements #5.

### CODE REVIEW

File: src/main/java/cs361/battleships/models/Board.java [hunk 1]
```diff
@@ -71,9 +75,57 @@ public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
...
+			res.setResult(AtackStatus.INVALID);
+			return res;
+		}
+
+		if (previouslyAttacked(x,y)){ // Invalid if the attack has already been attempted.
+			res.setResult(AtackStatus.INVALID);
+			return res;
+		}
+
+		if (!shipOnSpot(x,y)){ // If not a ship on the spot, record a miss.
```
Comment (lkthang) [inline src/main/java/cs361/battleships/models/Board.java]: so if there is a ship there, it returns 1? if it did hit then it wouldnt do the if statement and just do the "Ship attackedShip = hitShip(x,y) line?

File: src/main/java/cs361/battleships/models/Result.java [hunk 2]
```diff
@@ -2,30 +2,31 @@
 
 public class Result {
```
Comment (Bensonre) [inline src/main/java/cs361/battleships/models/Result.java]: Idk where you would add them but this class has no comments at all might be worth it to add some. even tho the code is straight forward

File: src/test/java/cs361/battleships/models/BoardTest.java [hunk 3]
```diff
@@ -49,16 +54,161 @@ public void testIfShipsOnBoard(){
...
+    }
+
+    @Test
+    public void testAttackAlreadyAttacked() {
+        Board board = new Board();
+
+        List<Result> attacks = new ArrayList<Result>();
+        Result adder = new Result();
+        adder.setResult(AtackStatus.MISS);
+        adder.setLocation(new Square(0,'A'));
```
Comment (Bensonre) [inline src/test/java/cs361/battleships/models/BoardTest.java]: Isn't row zero off the board? Would it be better to test this at 1,A?

File: src/main/java/cs361/battleships/models/Board.java [hunk 4]
```diff
@@ -71,9 +75,57 @@ public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
...
 	DO NOT change the signature of this method. It is used by the grading scripts.
 	 */
+
+	// Returns the result object of the attack AND adds it to the boards attacks.
 	public Result attack(int x, char y) {
-		//TODO Implement
-		return null;
+		Result res = new Result();
+
+		if (y > 'J' || y < 'A' || x > 10 || x < 0){ // Invalid if attack is off the board.
```
Comment (Bensonre) [inline src/main/java/cs361/battleships/models/Board.java]: shouldn't it be <= 0, I may be wrong but I believe row zero is off the board

Comment (xMestas) [reply src/main/java/cs361/battleships/models/Board.java]: Yes, good catch

File: src/main/java/cs361/battleships/models/Board.java (hunk 1 above)
Comment (xMestas) [reply src/main/java/cs361/battleships/models/Board.java]: shipOnSpot returns true if there is a ship on the spot.  The if statement says NOT ship on the spot, so it does the if statement if there is not a ship on the spot.  The if statement returns a MISS result after adding it to the attacks array.  If there is a ship on the spot, the function returns true, and the if statement is skipped.

File: src/main/java/cs361/battleships/models/Board.java [hunk 5]
```diff
@@ -85,11 +137,83 @@ public void setShips(List<Ship> ships) {
...
+		return false;
+	}
+
+	// Returns the ship being hit at that coordinate
+	public Ship hitShip(int x, char y){
+		for (int i = 0; i < ships.size(); i++){ // For all ships
+			List<Square> occupiedSquares = ships.get(i).getOccupiedSquares();  // Get ships squares
+			for (int j = 0; j < occupiedSquares.size(); j++){ // For each square
+				Square loc = occupiedSquares.get(j);
+				if (loc != null) {
```
[the remark this answers is not in the document]
Comment (xMestas) [reply src/main/java/cs361/battleships/models/Board.java]: yeet

File: src/main/java/cs361/battleships/models/Result.java (hunk 2 above)
Comment (xMestas) [reply src/main/java/cs361/battleships/models/Result.java]: Added comments for variables.

File: src/test/java/cs361/battleships/models/BoardTest.java (hunk 3 above)
Comment (xMestas) [reply src/test/java/cs361/battleships/models/BoardTest.java]: yup, you right.

**Pull request merged**